src/torchjd/autojac/mtl_backward.py

```diff
@@ -23,12 +25,92 @@
 )
 
 
+def _get_tensors_memory_id(tensor_list: list[Tensor]) -> set[int]:
+    """
+    Based on a list of lists of tensor objects returns the unique memory adress of all
+    :param tensor_list: list containing pytorch Tensor objects
+    """
+    return set([id(tensor) for tensor in tensor_list])
+
+
+def _traverse_ag_graph(grad_graph: Node) -> list[Tensor]:
+    """
+    Traverses an autograd graph iteratively and extracts all variables (trainable parameters)
+    that lead to the given node.
+
+    :param grad_graph: An autograd computation node.
+    :return: A list of tensors representing the variables that lead to the given node.
+    """
+    stack = [grad_graph]
+    output_params = []
+
+    while stack:
+        current_node = stack.pop()
+
+        child_nodes = current_node.next_functions
+        for child in child_nodes:
+            if child[0] is None:
+                # Skip constants in the calculation
+                continue
+            if hasattr(child[0], "variable"):
+                output_params.append(child[0].variable)
+            else:
+                stack.append(child[0])
+
+    return output_params
```

Comment:
I think this is wrong, for instance if `grad_graph` correspond to a leaf. I think this should fix it (and also improve readability)
```python
stack = [grad_graph]
output_params = []

while stack:
    current_node = stack.pop()
    
    if hasattr(current_node, "variable"):
        output_params.append(current_node)
    else:
        stack += [child[0] for child in current_node.next_functions if child[0] is not None]
return output_params
```

Reply:
The code breaks the tests. I also disagree here. If grad_graph is a leaf node the behaviour is correct as it returns an empty list here as there is no computation graph from which the leaf resulted. We can debate whether it makes sense to throw an error in this case instead as it is clearly a nonsensical input

Reply:
Well, I think that differentiating `x` with respect to `x` should be fine, right?

Reply:
Well, I think that differentiating `x` with respect to `x` should be fine, right?

Reply:
With the more recent implementation, the question becomes, when we do:
```python
a = torch.tensor(1.0, requires_grad=True, device=DEVICE)
leaves = _get_leaves_of_autograd_graph(tensors=[a], excluded=set())
```
should `leaves` be `{a}` or should it be `set()`?

Reply:
I settled for `set()` because when tensors do not have a `grad_fn`, they do not have an entry point to the autograd graph, so there is no "leaf" of the autograd graph.